File: matlab/drivers/runup_overtopping.m
clear all
clc

%% ACES Update to MATLAB
%-------------------------------------------------------------
% Driver for Wave Runup and Overtopping on Impermeable Structures (page 5-2
% in ACES User's Guide). Provides estimates of wave runup and overtopping
% on rough and smooth slope structures that are assumed to be impermeable

% Updated by: Mary Anderson, USACE-CHL-Coastal Processes Branch
% Date Created: April 18, 2011
% Date Verified: June 6, 2012 

% Requires the following functions:
% ERRSTP
% ERRWAVBRK2
% LWTDWS
% LWTGEN
% QOVERT
% QOVERT_IRR
% RUNUPR
% RUNUPS
% WAVELEN

% MAIN VARIABLE LIST:
%   MANDATORY INPUT
%   H: incident wave height (Hs for irregular waves)
%   T: wave period (Tp for irregular waves)
%   cotphi: cotan of nearshore slope
%   ds: water depth at structure toe
%   cottheta: cotan of structure slope (0.0 for vertical walls)
%   hs: structure height above toe 

%   OPTIONAL INPUT
%   a: empirical coefficient for rough slope runup
%   b: empirical coefficeint for rough slope runup
%   alpha: empirical coefficient for overtopping
%   Qstar0: empiricial coefficient for overtopping
%   U: onshore wind velocity for overtoppping
%   R: wave runup (if known)

%   MANDATORY OUTPUT
%   H0: deepwater wave height
%   relht0: deepwater realtive height (d/H)
%   steep0: deepwater wave steepness

%   OPTIONAL OUTPUT
%   R: wave runup
%   Q: overtopping

%   OTHERS
%   freeb: freeboard
%-------------------------------------------------------------

SET_PATHS();

[single_case] = USER_INPUT_SINGLE_MULTI_CASE();

[metric, g, labelUnitDist, labelUnitWt] = USER_INPUT_METRIC_IMPERIAL();

fprintf('%s \n\n','Calculation and slope type options: ');
fprintf('%s \n','Monochromatic Waves')
fprintf('%s \n','[1] Rough Slope <------------- Runup -------------> [2] Smooth Slope ')
fprintf('%s \n','[3] Rough Slope <----------- Overtopping ---------> [4] Smooth Slope ')
fprintf('%s \n\n','[5] Rough Slope <---- Runup and Overtopping ----> [6] Smooth Slope ')
fprintf('%s \n','Irregular Waves')
fprintf('%s \n\n','[7] Rough Slope <---- Runup and Overtopping ----> [8] Smooth Slope ')

option=input('Select option: ');

has_rough_slope = option==1 || option==5 || option ==7 || option==8;
has_overtopping = option>2;
has_runup = option ~= 3 && option ~= 4;
% numConsts = 0;
% if has_rough_slope
%     numConsts = numConsts + 2;
% end
% if has_overtopping
%     numConsts = numConsts + 3;
% end
% if ~has_runup
%     numConsts = numConsts + 1;
% end

conversionKnots2mph = 1.15077945; %1 knots = 1.15077945 mph

fprintf('\n');

% Single case input for metric measurments
if single_case
    [H] = USER_INPUT_DATA_VALUE(['Enter H: incident wave height (Hs for irregular waves) (' labelUnitDist '): '], 0.1, 100.0);
    
    [T] = USER_INPUT_DATA_VALUE(['Enter T: wave period (Tp for irregular waves) (' labelUnitDist '): '], 1.0, 1000.0);
    
    [cotphi] = USER_INPUT_DATA_VALUE('Enter cotphi: cotan of nearshore slope: ', 5.0, 10000.0);
    
    [ds] = USER_INPUT_DATA_VALUE(['Enter ds: water depth at structure toe (' labelUnitDist '): '], 0.1, 200.0);
    
    [cottheta] = USER_INPUT_DATA_VALUE('Enter cot \theta: cotan of structure slope (0.0 for vertical walls): ', 0.0, 30.0);
    
    [hs] = USER_INPUT_DATA_VALUE(['Enter hs: structure height above toe (' labelUnitDist '): '], 0.0, 200.0);

    numCases = 1;
    
else
     multiCaseData = {...
         ['H: wave height (' labelUnitDist ')'], 0.1, 100.0;...
          'T: wave period (sec)', 1.0, 1000.0;...
          'cotphi: cotan of nearshore slope', 5.0, 10000.0;...
         ['ds: water depth at structure toe (' labelUnitDist ')'], 0.1, 200.0;...
          'cottheta: cotan of structure slope (0.0 for vertical walls)', 0.0, 30.0;...
         ['hs: structure height above toe (' labelUnitDist ')'], 0.0, 200}; 
    
    [varData, numCases] = USER_INPUT_MULTI_MODE(multiCaseData);
    
    HList = varData(1, :);
    TList = varData(2, :);
    cotphiList = varData(3, :);
    dsList = varData(4, :);
    cotthetaList = varData(5, :);
    hsList = varData(6, :);
end

if option == 3
    R_default = 15.0;
elseif option == 4
    R_default = 20.0;
else
    R_default = 0;
end
[roughSlopeCoeffData] = USER_INPUT_ROUGH_SLOPE_COEFFICIENTS(...
    has_rough_slope, has_overtopping, has_runup,...
    struct('numCases', numCases, 'R_default', R_default));

if isfield(roughSlopeCoeffData, 'a')
    a = roughSlopeCoeffData.a;
end
if isfield(roughSlopeCoeffData, 'b')
    b = roughSlopeCoeffData.b;
end
if isfield(roughSlopeCoeffData, 'alpha')
    alpha = roughSlopeCoeffData.alpha;
end
if isfield(roughSlopeCoeffData, 'Qstar0')
    Qstar0 = roughSlopeCoeffData.Qstar0;
end
if isfield(roughSlopeCoeffData, 'U')
    U = roughSlopeCoeffData.U;
end
if isfield(roughSlopeCoeffData, 'R')
    R = roughSlopeCoeffData.R;
end

exporter = EXPORTER('output/exporterRunupOvertopping.txt');

% File Output
fileOutputArgs = {};
[fileOutputData] = USER_INPUT_FILE_OUTPUT(fileOutputArgs);

if fileOutputData{1}
    fId = fopen('output/runup_overtopping.txt', 'wt');
end

for loopIndex = 1:numCases
    if ~single_case
        H = HList(loopIndex);
        T = TList(loopIndex);
        cotphi = cotphiList(loopIndex);
        ds = dsList(loopIndex);
        cottheta = cotthetaList(loopIndex);
        hs = hsList(loopIndex);
        
        if isfield(roughSlopeCoeffData, 'aList')
            a = roughSlopeCoeffData.aList(loopIndex);
        end
        
        if isfield(roughSlopeCoeffData, 'bList')
            b = roughSlopeCoeffData.bList(loopIndex);
        end
        
        if isfield(roughSlopeCoeffData, 'alphaList')
            alpha = roughSlopeCoeffData.alphaList(loopIndex);
        end
        
        if isfield(roughSlopeCoeffData, 'Qstar0List')
            Qstar0 = roughSlopeCoeffData.Qstar0List(loopIndex);
        end
        
        if isfield(roughSlopeCoeffData, 'UList')
            U = roughSlopeCoeffData.UList(loopIndex);
        end
        
        if isfield(roughSlopeCoeffData, 'RList')
            R = roughSlopeCoeffData.RList(loopIndex);
        end
    end
    
    errorMsg = '';
    
    m=1/cotphi;

%     assert(ds<hs,'Error: Method does not apply to submerged structures.')
    if not(ds<hs)
        errorMsg = 'Error: Method does not apply to submerged structures.';
        disp(errorMsg);
    else
        [Hbs]=ERRWAVBRK2(T,m,ds);
%         assert(H<Hbs,'Error: Wave broken at structure (Hbs = %6.2f %s)',Hbs,labelUnitDist)
        if not (H<Hbs)
            errorMsg = sprintf('Error: Wave broken at structure (Hbs = %6.2f %s)',Hbs,labelUnitDist);
            disp(errorMsg);
        else
            [c,c0,cg,cg0,k,L,L0,reldep]=LWTGEN(ds,T,g);

            [steep,maxstp]=ERRSTP(H,ds,L);
%             assert(steep<maxstp,'Error: Input wave unstable (Max: %0.4f, [H/L] = %0.4f)',maxstp,steep')
            if not(steep<maxstp)
                errorMsg = sprintf('Error: Input wave unstable (Max: %0.4f, [H/L] = %0.4f)',maxstp,steep');
                disp(errorMsg);
            else
                [alpha0,H0]=LWTDWS(0,c,cg,c0,H);

                relht0=ds/H0;
                steep0=H0/(g*T^2);

                if cottheta==0
%                     assert(option~=1 && option~=5 && option~=7,'Error: Vertical wall cannot have rough slope.')
                    if not(option~=1 && option~=5 && option~=7)
                        errorMsg = 'Error: Vertical wall cannot have rough slope.';
                        disp(errorMsg);
                    else
                        theta=0.5*pi;
                        ssp=1000;
                    end
                else
                    theta=atan(1/cottheta);
                    ssp=(1/cottheta)/sqrt(H/L0);
                end

                if length(errorMsg) == 0
                    fprintf('%s \n','Deepwater')
                    fprintf('\t %s \t\t\t\t %-6.3f %s \n','Wave height, Hs0',H0,labelUnitDist)
                    fprintf('\t %s \t\t %-6.3f \n','Relative height, ds/H0',relht0)
                    fprintf('\t %s \t %-6.6f \n\n','Wave steepness, Hs0/(gT^2)',steep0)
                end
            end
        end
    end

    if fileOutputData{1}
        if ~single_case
            fprintf(fId, 'Case #%d\n\n', loopIndex);
        end
        
        fprintf(fId, 'Input\n');
        fprintf(fId, 'H\t\t%6.2f %s\n', H, labelUnitDist);
        fprintf(fId, 'T\t\t%6.2f s\n', T);
        fprintf(fId, 'cotphi\t\t%6.2f\n', cotphi);
        fprintf(fId, 'ds\t\t%6.2f %s\n', ds, labelUnitDist);
        fprintf(fId, 'cottheta\t%6.2f\n', cottheta);
        fprintf(fId, 'hs\t\t%6.2f %s\n', hs, labelUnitDist);
        
        exportData = {H, T, cotphi, ds, cottheta, hs};
        
        if exist('a') == 1
            fprintf(fId, 'a\t\t%6.4f\n', a);
            exportData = [exportData {a}];
        end
        
        if exist('b') == 1
            fprintf(fId, 'b\t\t%6.4f\n', b);
            exportData = [exportData {b}];
        end
        
        if exist('alpha') == 1
            fprintf(fId, 'alpha\t\t%6.4f\n', alpha);
            exportData = [exportData {alpha}];
        end
        
        if exist('Qstar0') == 1
            fprintf(fId, 'Qstar0\t\t%6.4f\n', Qstar0);
            exportData = [exportData {Qstar0}];
        end
        
        if exist('U') == 1
            fprintf(fId, 'U\t\t%6.4f knots\n', U/conversionKnots2mph);
            exportData = [exportData {U/conversionKnots2mph}];
        end
        
        if exist('R') == 1
            fprintf(fId, 'R\t\t%6.4f %s\n', R, labelUnitDist);
            if ~has_runup
                exportData = [exportData {R}];
            end
        end
        
        if length(errorMsg) > 0
            fprintf(fId, '\n%s\n\n', errorMsg);
            exportData = [exportData {errorMsg}];
        else
            fprintf(fId, '\n%s \n','Deepwater');
            fprintf(fId, '\t %s \t\t %6.3f %s \n','Wave height, Hs0',H0,labelUnitDist);
            fprintf(fId, '\t %s \t %6.3f \n','Relative height, ds/H0',relht0);
            fprintf(fId, '\t %s \t %6.6f \n\n','Wave steepness, Hs0/(gT^2)',steep0);
            
            exportData = [exportData {H0, relht0, steep0}];
        end
    end
    
    if length(errorMsg) == 0
        freeb=hs-ds;

        if option==1
            [R]=RUNUPR(H,ssp,a,b);
            fprintf('%s \t %-6.3f %s \n\n','Runup',R,labelUnitDist)

            if fileOutputData{1}
                fprintf(fId, '%s \t %6.3f %s \n\n','Runup',R,labelUnitDist);
                exportData = [exportData {R}];
            end
        elseif option==2
            [R]=RUNUPS(H,L,ds,theta,ssp);
            fprintf('%s \t %-6.3f %s \n\n','Runup',R,labelUnitDist)

            if fileOutputData{1}
                fprintf(fId, '%s \t %6.3f %s \n\n','Runup',R,labelUnitDist);
                exportData = [exportData {R}];
            end
        elseif option==3   
            [Q]=QOVERT(H0,freeb,R,Qstar0,alpha,theta,U,g);
            fprintf('%s \t %-6.3f %s^3/sec-%s \n\n','Overtopping rate per unit width',Q,labelUnitDist,labelUnitDist)

            if fileOutputData{1}
                fprintf(fId, '%s \t %6.3f %s^3/sec-%s \n\n','Overtopping rate per unit width',Q,labelUnitDist,labelUnitDist);
                exportData = [exportData {Q}];
            end
        elseif option==4
            [Q]=QOVERT(H0,freeb,R,Qstar0,alpha,theta,U,g);
            fprintf('%s \t %-6.3f %s^3/sec-%s \n\n','Overtopping rate per unit width',Q,labelUnitDist,labelUnitDist)

            if fileOutputData{1}
                fprintf(fId, '%s \t %6.3f %s^3/sec-%s \n\n','Overtopping rate per unit width',Q,labelUnitDist,labelUnitDist);
                exportData = [exportData {Q}];
            end
        elseif option==5
            [R]=RUNUPR(H,ssp,a,b);
            [Q]=QOVERT(H0,freeb,R,Qstar0,alpha,theta,U,g);
            fprintf('%s \t %-6.3f %s \n','Runup',R,labelUnitDist)
            fprintf('%s \t %-6.3f %s^3/sec-%s \n\n','Overtopping rate per unit width',Q,labelUnitDist,labelUnitDist)

            if fileOutputData{1}
                fprintf(fId, '%s \t %6.3f %s \n','Runup',R,labelUnitDist);
                fprintf(fId, '%s \t %6.3f %s^3/sec-%s \n\n','Overtopping rate per unit width',Q,labelUnitDist,labelUnitDist);
                exportData = [exportData {R, Q}];
            end
        elseif option==6
            [R]=RUNUPS(H,L,ds,theta,ssp);
            [Q]=QOVERT(H0,freeb,R,Qstar0,alpha,theta,U,g);
            fprintf('%s \t %-6.3f %s \n','Runup',R,labelUnitDist)
            fprintf('%s \t %-6.3f %s^3/sec-%s \n\n','Overtopping rate per unit width',Q,labelUnitDist,labelUnitDist)

            if fileOutputData{1}
                fprintf(fId, '%s \t %6.3f %s \n','Runup',R,labelUnitDist);
                fprintf(fId, '%s \t %6.3f %s^3/sec-%s \n\n','Overtopping rate per unit width',Q,labelUnitDist,labelUnitDist);
                exportData = [exportData {R, Q}];
            end
        elseif option==7
            [R]=RUNUPR(H,ssp,a,b);
            [Q]=QOVERT_IRR(H0,freeb,R,Qstar0,alpha,theta,U,g);
            fprintf('%s \t %-6.3f %s \n','Runup',R,labelUnitDist)
            fprintf('%s \t %-6.3f %s^3/sec-%s \n\n','Overtopping rate per unit width',Q,labelUnitDist,labelUnitDist)

            if fileOutputData{1}
                fprintf(fId, '%s \t %6.3f %s \n','Runup',R,labelUnitDist);
                fprintf(fId, '%s \t %6.3f %s^3/sec-%s \n\n','Overtopping rate per unit width',Q,labelUnitDist,labelUnitDist);
                exportData = [exportData {R, Q}];
            end
        elseif option==8
            [R]=RUNUPS(H,L,ds,theta,ssp);
            [Q]=QOVERT_IRR(H0,freeb,R,Qstar0,alpha,theta,U,g);
            fprintf('%s \t %-6.3f %s \n','Runup',R,labelUnitDist)
            fprintf('%s \t %-6.3f %s^3/sec-%s \n\n','Overtopping rate per unit width',Q,labelUnitDist,labelUnitDist)

            if fileOutputData{1}
                fprintf(fId, '%s \t %6.3f %s \n','Runup',R,labelUnitDist);
                fprintf(fId, '%s \t %6.3f %s^3/sec-%s \n\n','Overtopping rate per unit width',Q,labelUnitDist,labelUnitDist);
                exportData = [exportData {R, Q}];
            end
        end
    end
    
    if fileOutputData{1}
        if loopIndex < numCases
            fprintf(fId, '\n--------------------------------------\n\n');
        end
        
        exporter.writeData(exportData);
    end
end

if fileOutputData{1}
    fclose(fId);
end

exporter.close();

File: python/drivers/runup_overtopping.py
import sys
import math
import numpy as np
sys.path.append('../functions')

from base_driver import BaseDriver
from helper_objects import BaseField
import USER_INPUT
from ERRSTP import ERRSTP
from ERRWAVBRK2 import ERRWAVBRK2
from LWTDWS import LWTDWS
from LWTGEN import LWTGEN
from QOVERT import QOVERT
from QOVERT_IRR import QOVERT_IRR
from RUNUPR import RUNUPR
from RUNUPS import RUNUPS
from WAVELEN import WAVELEN

#check for option 3, vert wall
## ACES Update to MATLAB
#-------------------------------------------------------------
# Driver for Wave Runup and Overtopping on Impermeable Structures (page 5-2
# in ACES User's Guide). Provides estimates of wave runup and overtopping
# on rough and smooth slope structures that are assumed to be impermeable

# Updated by: Mary Anderson, USACE-CHL-Coastal Processes Branch
# Date Created: April 18, 2011
# Date Verified: June 6, 2012

# Requires the following functions:
# ERRSTP
# ERRWAVBRK2
# LWTDWS
# LWTGEN
# QOVERT
# QOVERT_IRR
# RUNUPR
# RUNUPS
# WAVELEN

# MAIN VARIABLE LIST:
#   MANDATORY INPUT
#   H: incident wave height (Hs for irregular waves)
#   T: wave period (Tp for irregular waves)
#   cotphi: cotan of nearshore slope
#   ds: water depth at structure toe
#   cottheta: cotan of structure slope (0.0 for vertical walls)
#   hs: structure height above toe

#   OPTIONAL INPUT
#   a: empirical coefficient for rough slope runup
#   b: empirical coefficeint for rough slope runup
#   alpha: empirical coefficient for overtopping
#   Qstar0: empiricial coefficient for overtopping
#   U: onshore wind velocity for overtoppping
#   R: wave runup (if known)

#   MANDATORY OUTPUT
#   H0: deepwater wave height
#   relht0: deepwater realtive height (d / H)
#   steep0: deepwater wave steepness

#   OPTIONAL OUTPUT
#   R: wave runup
#   Q: overtopping

#   OTHERS
#   freeb: freeboard
#-------------------------------------------------------------

class RunupOvertopping(BaseDriver):
    def __init__(self, H = None, T = None, cotphi = None,\
        ds = None, cottheta = None, hs = None):
        if H != None:
            self.isSingleCase = True
            self.defaultValueH = H
        if T != None:
            self.isSingleCase = True
            self.defaultValueT = T
        if cotphi != None:
            self.isSingleCase = True
            self.defaultValue_cotphi = cotphi
        if ds != None:
            self.isSingleCase = True
            self.defaultValue_ds = ds
        if cottheta != None:
            self.isSingleCase = True
            self.defaultValue_cottheta = cottheta
        if hs != None:
            self.isSingleCase = True
            self.defaultValue_hs = hs

        self.conversionKnots2mph = 1.15077945 #1 knots = 1.15077945 mph

        super(RunupOvertopping, self).__init__()
    # end __init__

    def userInput(self):
        print('%s \n\n' % 'Calculation and slope type options: ')
        print('%s \n' % 'Monochromatic Waves')
        print('%s \n' % '[1] Rough <------------- Runup -------------> [2] Smooth')
        print('%s \n' % '[3] Rough <----------- Overtopping ---------> [4] Smooth')
        print('%s \n\n' % '[5] Rough <----- Runup and Overtopping -----> [6] Smooth')
        print('%s \n' % 'Irregular Waves')
        print('%s \n\n' % '[7] Rough <----- Runup and Overtopping -----> [8] Smooth')

        self.option = USER_INPUT.FINITE_CHOICE("Select option (1 - 8): ",\
            ["1", "2", "3", "4", "5", "6", "7", "8"])
        self.option = int(self.option)

        self.has_rough_slope = self.option == 1 or\
            self.option == 5 or self.option == 7 or self.option == 8
        self.has_overtopping = self.option > 2
        self.has_runup = self.option != 3 and self.option != 4

        if self.option == 3:
            R_default = 15.0
        elif self.option == 4:
            R_default = 20.0
        else:
            R_default = 0.0

        super(RunupOvertopping, self).userInput()

        print(self.inputList)
        # self.roughSlopeCoeffDict = {"a": 0.956, "b": 0.398,\
        #     "alpha": 0.076473, "Qstar0": 0.025, "U": 35.0, "R": R_default}
        if self.option != 2:
            self.roughSlopeCoeffDict = USER_INPUT.ROUGH_SLOPE_COEFFICIENTS(\
                self.has_rough_slope, self.has_overtopping, self.has_runup,\
                {"numCases": len(self.dataOutputList), "R_default": R_default})
        else:
            self.roughSlopeCoeffDict = {}
    # end userInput

    def defineInputDataList(self):
        self.inputList = []

        if not hasattr(self, "defaultValueH"):
            self.inputList.append(BaseField("H: incident wave height (Hs for irregular waves) (%s)" %\
                (self.labelUnitDist), 0.1, 100.0))
        if not hasattr(self, "defaultValueT"):
            self.inputList.append(BaseField("T: wave period (Tp for irregular waves) (s)", 1.0, 1000.0))
        if not hasattr(self, "defaultValue_cotphi"):
            self.inputList.append(BaseField("cotphi: cotan of nearshore slope", 5.0, 10000.0))
        if not hasattr(self, "defaultValue_ds"):
            self.inputList.append(BaseField("ds: water depth at structure toe (%s)" %\
                self.labelUnitDist, 0.1, 200.0))
        if not hasattr(self, "defaultValue_cottheta"):
            self.inputList.append(BaseField("cottheta: cotan of structure slope (0.0 for vertical walls)",\
                0.0, 30.0))
        if not hasattr(self, "defaultValue_hs"):
            self.inputList.append(BaseField("hs: structure height above toe (%s)" % self.labelUnitDist,\
                0.0, 200.0))
    # end defineInputDataList

    def fileOutputRequestInit(self):
        self.fileOutputRequestMain(defaultFilename = "runup_overtopping")

    def getCalcValues(self, caseInputList, caseIndex):
        currIndex = 0

        if hasattr(self, "defaultValueH"):
            H = self.defaultValueH
        else:
            H = caseInputList[currIndex]
            currIndex = currIndex + 1

        if hasattr(self, "defaultValueT"):
            T = self.defaultValueT
        else:
            T = caseInputList[currIndex]
            currIndex = currIndex + 1

        if hasattr(self, "defaultValue_cotphi"):
            cotphi = self.defaultValue_cotphi
        else:
            cotphi = caseInputList[currIndex]
            currIndex = currIndex + 1

        if hasattr(self, "defaultValue_ds"):
            ds = self.defaultValue_ds
        else:
            ds = caseInputList[currIndex]
            currIndex = currIndex + 1

        if hasattr(self, "defaultValue_cottheta"):
            cottheta = self.defaultValue_cottheta
        else:
            cottheta = caseInputList[currIndex]
            currIndex = currIndex + 1

        if hasattr(self, "defaultValue_hs"):
            hs = self.defaultValue_hs
        else:
            hs = caseInputList[currIndex]

        print(self.roughSlopeCoeffDict)

        coeffDict = {}
        for coeffName in ["a", "b", "alpha", "Qstar0", "U", "R"]:
            if coeffName in self.roughSlopeCoeffDict:
                coeffDict[coeffName] = self.roughSlopeCoeffDict[coeffName]
            elif (coeffName + "List") in self.roughSlopeCoeffDict:
                coeffDict[coeffName] =\
                    self.roughSlopeCoeffDict[coeffName + "List"][caseIndex]
            else:
                coeffDict[coeffName] = None

        return H, T, cotphi, ds, cottheta, hs,\
            coeffDict["a"],\
            coeffDict["b"],\
            coeffDict["alpha"],\
            coeffDict["Qstar0"],\
            coeffDict["U"],\
            coeffDict["R"]
    # end getCalcValues

    def performCalculations(self, caseInputList, caseIndex = 0):
        H, T, cotphi, ds, cottheta, hs, a, b, alpha, Qstar0, U, R =\
            self.getCalcValues(caseInputList, caseIndex)

        m = 1.0/cotphi

        if not (ds < hs):
            print("Error: Method does not apply to submerged structures.")
            return

        Hbs = ERRWAVBRK2(T, m, ds)
        if not (H < Hbs):
            print("Error: Wave broken at structure (Hbs = %6.2f %s)" %\
                (Hbs, self.labelUnitDist))
            return

        c, c0, cg, cg0, k, L, L0, reldep = LWTGEN(ds, T, self.g)

        steep, maxstp = ERRSTP(H, ds, L)
        if not (steep < maxstp):
            print("Error: Input wave unstable (Max: %0.4f. [H/L] = %0.4f" %\
                (maxstp, steep))
            return

        alpha0, H0 = LWTDWS(0.0, c, cg, c0, H)

        relht0 = ds/H0
        steep0 = H0 / (self.g * T**2)

        if np.isclose(cottheta, 0):
            if not (self.option != 1):
                print("Error: Vertical wall cannot have rough slope.")
                return

            theta = 0.5*math.pi
            ssp = 1000
        else:
            theta = math.atan(1.0/cottheta)
            ssp = (1.0/cottheta)/math.sqrt(H/L0)
        # end if

        print("Deepwater")
        print("\tWave height, Hs0\t\t%-6.3f %s" % (H0, self.labelUnitDist))
        print("\tRelative height, ds/H0\t\t%-6.3f" % relht0)
        print("\tWave steepness, Hs0/(gT^2)\t%-6.6f" % steep0)

        freeb = hs - ds

        dataDict = {"H": H, "T": T, "cotphi": cotphi, "ds": ds,\
            "cottheta": cottheta, "hs": hs, "a": a, "b": b,\
            "alpha": alpha, "Qstar0": Qstar0, "U": U, "H0": H0,\
            "relht0": relht0, "steep0": steep0}

        if self.option == 1:
            R = RUNUPR(H, ssp, a, b)
        elif self.option == 2:
            R = RUNUPS(H, L, ds, theta, ssp)
        elif self.option == 3:
            Q = QOVERT(H0, freeb, R, Qstar0, alpha, theta, U, self.g)
        elif self.option == 4:
            Q = QOVERT(H0, freeb, R, Qstar0, alpha, theta, U, self.g)
        elif self.option == 5:
            R = RUNUPR(H, ssp, a, b)
            Q = QOVERT(H0, freeb, R, Qstar0, alpha, theta, U, self.g)
        elif self.option == 6:
            R = RUNUPS(H, L, ds, theta, ssp)
            Q = QOVERT(H0, freeb, R, Qstar0, alpha, theta, U, self.g)
        elif self.option == 7:
            R = RUNUPR(H, ssp, a, b)
            Q = QOVERT_IRR(H0, freeb, R, Qstar0, alpha, theta, U, self.g)
        elif self.option == 8:
            R = RUNUPS(H, L, ds, theta, ssp)
            Q = QOVERT_IRR(H0, freeb, R, Qstar0, alpha, theta, U, self.g)

        if self.option == 1 or self.option == 2 or\
            self.option == 5 or self.option == 6 or\
            self.option == 7 or self.option == 8:
            print("Runup\t\t\t\t\t%-6.3f %s" % (R, self.labelUnitDist))
        if self.option == 3 or self.option == 4 or\
            self.option == 5 or self.option == 6 or\
            self.option == 7 or self.option == 8:
            print("Overtopping rate per unit width\t\t%-6.3f %s^3/sec-%s" %\
                (Q, self.labelUnitDist, self.labelUnitDist))
            dataDict["Q"] = Q

        dataDict["R"] = R
        self.fileOutputWriteMain(dataDict, caseIndex)
    # end performCalculations

    def fileOutputWriteData(self, dataDict):
        self.fileRef.write("Input\n")
        self.fileRef.write("H\t\t%6.2f %s\n" %\
            (dataDict["H"], self.labelUnitDist))
        self.fileRef.write("T\t\t%6.2f s\n" % dataDict["T"])
        self.fileRef.write("cotphi\t\t%6.2f\n" % dataDict["cotphi"])
        self.fileRef.write("ds\t\t%6.2f %s\n" %\
            (dataDict["ds"], self.labelUnitDist))
        self.fileRef.write("cottheta\t%6.2f\n" % dataDict["cottheta"])
        self.fileRef.write("hs\t\t%6.2f %s\n" %\
            (dataDict["hs"], self.labelUnitDist))

        if "a" in dataDict and dataDict["a"] != None:
            self.fileRef.write("a\t\t%6.4f\n" % dataDict["a"])
        if "b" in dataDict and dataDict["b"] != None:
            self.fileRef.write("b\t\t%6.4f\n" % dataDict["b"])
        if "alpha" in dataDict and dataDict["alpha"] != None:
            self.fileRef.write("alpha\t\t%6.4f\n" % dataDict["alpha"])
        if "Qstar0" in dataDict and dataDict["Qstar0"] != None:
            self.fileRef.write("Qstar0\t\t%6.4f\n" % dataDict["Qstar0"])
        if "U" in dataDict and dataDict["U"] != None:
            self.fileRef.write("U\t\t%6.4f kn\n" %\
                (dataDict["U"]/self.conversionKnots2mph))
        if "R" in dataDict and dataDict["R"] != None and not self.has_runup:
            self.fileRef.write("R\t\t%6.4f %s\n" %\
                (dataDict["R"], self.labelUnitDist))

        self.fileRef.write("\nDeepwater\n")
        self.fileRef.write("\tWave height, Hs0\t\t%-6.3f %s\n" %\
            (dataDict["H0"], self.labelUnitDist))
        self.fileRef.write("\tRelative height, ds/H0\t\t%-6.3f\n" %\
            dataDict["relht0"])
        self.fileRef.write("\tWave steepness, Hs0/(gT^2)\t%-6.6f\n" %\
            dataDict["steep0"])

        if self.option == 1 or self.option == 2 or\
            self.option == 5 or self.option == 6 or\
            self.option == 7 or self.option == 8:
            self.fileRef.write("Runup\t\t\t\t\t%-6.3f %s\n" %\
                (dataDict["R"], self.labelUnitDist))
        if self.option == 3 or self.option == 4 or\
            self.option == 5 or self.option == 6 or\
            self.option == 7 or self.option == 8:
            self.fileRef.write("Overtopping rate per unit width\t\t%-6.3f %s^3/sec-%s\n" %\
                (dataDict["Q"], self.labelUnitDist, self.labelUnitDist))
    # end fileOutputWriteData


driver = RunupOvertopping()